keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,1767 @@
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+In certain specific scenarios, there are risks associated with enabling deferral
+of file system operations from CSI plugins to container runtime handlers.
+Careful planning is necessa and mitigations for these are enumerated below.
+
+#### Failures after CSI plugin downgrade
+
+If the CSI node plugin is downgraded with respect to it's ability to defer
+file system operations to a container runtime, CSI operations like obtaining
+stats and online expansion of volumes mounted by the container runtime will
+start to fail. The errors should be clearly surfaced to the user (by Kubelet)
+thorugh events on the volume. Recovering from this will require the pod to be
+killed so that the volume gets re-published.
+
+Graceful termination of the pod should work out in most scenarios as the
+downgraded CSI plugin will find it has no mounts to clean up and return success
+(in adherence with the idempotent requirements of multiple invocations of CSI
+APIs). To avoid any failure during NodeUnpublishVolume or leakage of metadata
+files (populated earlier by the CSI plugin in accordance with the file system
+deferral strategy before the downgrade affecting the ability to defer file
+system operations to the container runtime) that are not automatically cleaned
+up the runtime, a node should be drained and cordoned before downgrading the CSI
+node plugin across versions with different capabilities around deferral of file
+system operations to a container runtime handler.
+
+#### Potential data corruption for co-scheduled pods accessing the same PVC
+
+If a pod controller launches multiple pods that are co-scheduled on the same
+node and refer to the same PVC with access mode set to ReadWriteOnce and bound
+to a volume backed by a block device with a file system that does not support
+multiple, simultaneous read-write mounts, there may be data corruption when
+switching to deffered mounting of file systems by a microvm runtime like Kata.
+This results from the fact that each VM (running each co-scheduled pod mounting
+the same PVC) will try to mount the same file system multiple times within the
+VM. Thus, this effectively switches the access mode for the PVC from
+ReadWriteOnce to ReadWriteMany which can lead to corruption.
+
+To avoid such corruption, there are several options:
+- Introduce safety checks in a CSI plugin (that has the capability to defer file
+system operations to a container runtime) to fail a NodePublishVolume call if
+the volume is already found to be published on the same node to a different
+target path and the file system cannot handle multiple simultaneous mounts.
```

Comment:
This will break existing applications that use ReadWriteOnce volumes to run two pods on the same node with it.

Reply:
Removed the above recommendation.